// extract-first-json/build/index.js
"use strict";

exports.__esModule = true;
exports.extractJSONObject = extractJSONObject;
exports.extractJSONArray = extractJSONArray;
exports.extractJSON = extractJSON;

var _parseJsonObject = require("parse-json-object");

exports.JSONObject = _parseJsonObject.JSONObject;
exports.JSONArray = _parseJsonObject.JSONArray;

var _reduceFirst = _interopRequireDefault(require("reduce-first"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function extract(text, startBracket, endBracket, parser) {
  return (0, _reduceFirst["default"])(Array.from(text), function (_char, startIndex) {
    if (_char === startBracket) {
      var section = text.slice(startIndex);

      while (section.lastIndexOf(endBracket) > 0) {
        section = section.slice(0, section.lastIndexOf(endBracket) + 1);
        var result = parser(section);

        if (result !== undefined) {
          return result;
        }
      }
    }

    return undefined;
  });
}

function extractJSONObject(text) {
  return extract(text, "{", "}", _parseJsonObject.parseJSONObject);
}

function extractJSONArray(text) {
  return extract(text, "[", "]", _parseJsonObject.parseJSONArray);
}

function extractJSON(text) {
  var _extractJSONObject;

  return (_extractJSONObject = extractJSONObject(text)) != null ? _extractJSONObject : extractJSONArray(text);
}
